File: src/main/java/com/technews/handler/RestControllerExceptionHandler.java
package com.technews.handler;

import com.technews.common.dto.BasicResponse;
import com.technews.common.exception.FirebaseQuerySnapshotException;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.BeanCreationNotAllowedException;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.http.converter.HttpMessageNotReadableException;
import org.springframework.web.bind.MissingServletRequestParameterException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.bind.annotation.RestControllerAdvice;
import org.springframework.web.method.annotation.MethodArgumentTypeMismatchException;
import org.springframework.web.servlet.NoHandlerFoundException;

import java.net.BindException;
import java.nio.file.AccessDeniedException;

@Slf4j
@RestControllerAdvice(annotations = RestController.class)
public class RestControllerExceptionHandler {

    @ExceptionHandler({
            HttpMessageNotReadableException.class,
            MethodArgumentTypeMismatchException.class,
            BindException.class,
            IllegalArgumentException.class,
            MissingServletRequestParameterException.class,
            AccessDeniedException.class
    })
    public ResponseEntity handleBadRequest(Exception ex) {
        return BasicResponse.clientError(ex.getMessage());
    }

    @ExceptionHandler({
            FirebaseQuerySnapshotException.class,
    })
    public ResponseEntity handleCustomNotFoundException(Exception ex) {
        return BasicResponse.internalServerError(ex.getMessage());
    }

    @ExceptionHandler(Exception.class)
    public ResponseEntity handleException(Exception ex) {
        log.error("Exception. ", ex);
        return BasicResponse.internalServerError("An error occurred while processing the request. Please try again.");
    }

    @ExceptionHandler(BeanCreationNotAllowedException.class)
    public ResponseEntity handleBeanCreationNotAllowedException(Exception ex) {
        log.error("BeanCreationNotAllowedException. ", ex);
        return new ResponseEntity("BeanCreationNotAllowedException", HttpStatus.SERVICE_UNAVAILABLE);
    }

    @ExceptionHandler(NoHandlerFoundException.class)
    public ResponseEntity handleNoHandlerFoundException(Exception ex) {
        log.error("NoHandlerFoundException. ", ex);
        return BasicResponse.notFound("NoHandlerFoundException");
    }
}

File: src/main/kotlin/com/technews/handler/RestControllerExceptionHandler.kt
package com.technews.handler

import com.technews.common.constant.Result
import com.technews.common.dto.BasicResponse
import mu.KotlinLogging
import org.springframework.beans.factory.BeanCreationNotAllowedException
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.http.converter.HttpMessageNotReadableException
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.MissingServletRequestParameterException
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestController
import org.springframework.web.bind.annotation.RestControllerAdvice
import org.springframework.web.method.annotation.MethodArgumentTypeMismatchException
import org.springframework.web.servlet.NoHandlerFoundException
import java.net.BindException
import java.nio.file.AccessDeniedException

private val logger = KotlinLogging.logger {}

@RestControllerAdvice(annotations = [RestController::class])
class RestControllerExceptionHandler {
    @ExceptionHandler(MethodArgumentNotValidException::class)
    fun handleValidationExceptions(ex: MethodArgumentNotValidException): ResponseEntity<*> {
        val errorMessage = ex.bindingResult.allErrors.firstOrNull()?.defaultMessage ?: Result.FAIL.message
        return BasicResponse.clientError(errorMessage)
    }

    @ExceptionHandler(
        HttpMessageNotReadableException::class,
        MethodArgumentTypeMismatchException::class,
        BindException::class,
        IllegalArgumentException::class,
        MissingServletRequestParameterException::class,
        AccessDeniedException::class,
    )
    fun handleBadRequest(ex: Exception): ResponseEntity<*> =
        BasicResponse.clientError(ex.message.orEmpty())

    @ExceptionHandler(Exception::class)
    fun handleException(ex: Exception): ResponseEntity<*> {
        logger.error("Exception. ", ex)
        return BasicResponse.internalServerError("An error occurred while processing the request. Please try again.")
    }

    @ExceptionHandler(BeanCreationNotAllowedException::class)
    fun handleBeanCreationNotAllowedException(ex: Exception): ResponseEntity<*> {
        logger.error("BeanCreationNotAllowedException. ", ex)
        return ResponseEntity<Any>("BeanCreationNotAllowedException", HttpStatus.SERVICE_UNAVAILABLE)
    }

    @ExceptionHandler(NoHandlerFoundException::class)
    fun handleNoHandlerFoundException(ex: Exception): ResponseEntity<*> {
        logger.error("NoHandlerFoundException.", ex)
        return BasicResponse.notFound("NoHandlerFoundException")
    }
}
